single_kernel_mongo/managers/k8s.py

```diff
@@ -0,0 +1,271 @@
+from lightkube.resources.core_v1 import Node, Pod, Service
+
+from single_kernel_mongo.exceptions import FailedToFindNodePortError, FailedToFindServiceError
+
+# default logging from lightkube httpx requests is very noisy
+logging.getLogger("lightkube").disabled = True
+logging.getLogger("lightkube.core.client").disabled = True
+logging.getLogger("httpx").disabled = True
+logging.getLogger("httpcore").disabled = True
+
+logger = logging.getLogger()
+
+
+class K8sManager:
+    """Manager for handling k8s resources for mongo charms."""
+
+    def __init__(self, pod_name: str, namespace: str):
+        self.pod_name: str = pod_name
+        self.app_name: str = "-".join(pod_name.split("-")[:-1])
+        self.namespace: str = namespace
+
+    def __eq__(self, other: object) -> bool:
+        """__eq__ dunder.
+
+        Allows to get cache hit on calls on the same method from different instances of K8sManager
+        as `self` is passed to methods.
+        """
+        return isinstance(other, K8sManager) and self.__dict__ == other.__dict__
+
+    def __hash__(self) -> int:
+        """__hash__ dunder.
+
+        For dict like caching.
+        """
+        return hash(json.dumps(self.__dict__, sort_keys=True))
+
+    @cached_property
+    def client(self) -> Client:
+        """The Lightkube client."""
+        return Client(  # pyright: ignore[reportArgumentType]
+            field_manager=self.pod_name,
+            namespace=self.namespace,
+        )
+
+    # BEGIN: getters
+    def get_ttl_hash(self, seconds=120) -> int:
+        """Gets a unique time hash for the cache, expiring after 2 minutes.
+
+        We enforce a cache miss by using a ghost argument which changes every 2
+        minutes to all getters.
+        """
+        return math.floor(time.time() / seconds)
+
+    def get_pod(self, pod_name: str = "") -> Pod:
+        """Gets the pod via k8s API."""
+        return self._get_pod(pod_name, self.get_ttl_hash())
+
+    def get_node(self, pod_name: str) -> Node:
+        """Gets the node the port is running on."""
+        return self._get_node(pod_name, self.get_ttl_hash())
+
+    def get_node_ip(self, pod_name: str) -> str:
+        """Gets the IP Address of the Node via the K8s API."""
+        return self._get_node_ip(pod_name, self.get_ttl_hash())
+
+    def get_service(self, service_name: str) -> Service | None:
+        """Gets the Service via the K8s API."""
+        return self._get_service(service_name, self.get_ttl_hash())
+
+    def get_partition(self) -> int:
+        """Gets the stateful set rolling partition."""
+        return self._get_partition(self.get_ttl_hash())
+
+    def get_unit_service_name(self, pod_name: str = "") -> str:
+        """Returns the service name for the current unit."""
+        pod_name = pod_name or self.pod_name
+        return f"{pod_name}-external"
+
+    # END: getters
+
+    # BEGIN: helpers
+
+    def on_deployed_without_trust(self) -> None:
```

Comment:
nit: might make more sense to name `log_deployed_without_trust`

Reply:
There's something I'm trying so see and iron out here:
I would like to say "Hey, you're on k8s and you deployed without trust, I'm going to error state".
And there are some ways that make it easy to do (give the substrate to the manager), but I'd love to find a better solution and for now I don't have one :)

Reply:
If that is the case can you add a TODO here with all of that context?

Reply:
I have done it in the end, so resolving this now